graphistry/scrape_utils.py

```diff
@@ -0,0 +1,357 @@
+import hashlib
+import re
+import requests
+import os
+import json
+from urllib.parse import urlparse
+from collections import Counter
+import pandas as pd
+
+import tempfile
+from datetime import datetime
+
+# pip3 install newspaper3k
+import newspaper
+
+# pip install textract
+import textract
+
+from graphistry.util import setup_logger
+from graphistry.constants import VERBOSE, TRACE
+
+logger = setup_logger("scrape_utils", verbose=VERBOSE, fullpath=TRACE)
+
+URL_REGEX = r"\b((?:https?://)?(?:(?:www\.)?(?:[\da-z\.-]+)\.(?:[a-z]{2,6})|(?:(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.){3}(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)|(?:(?:[0-9a-fA-F]{1,4}:){7,7}[0-9a-fA-F]{1,4}|(?:[0-9a-fA-F]{1,4}:){1,7}:|(?:[0-9a-fA-F]{1,4}:){1,6}:[0-9a-fA-F]{1,4}|(?:[0-9a-fA-F]{1,4}:){1,5}(?::[0-9a-fA-F]{1,4}){1,2}|(?:[0-9a-fA-F]{1,4}:){1,4}(?::[0-9a-fA-F]{1,4}){1,3}|(?:[0-9a-fA-F]{1,4}:){1,3}(?::[0-9a-fA-F]{1,4}){1,4}|(?:[0-9a-fA-F]{1,4}:){1,2}(?::[0-9a-fA-F]{1,4}){1,5}|[0-9a-fA-F]{1,4}:(?:(?::[0-9a-fA-F]{1,4}){1,6})|:(?:(?::[0-9a-fA-F]{1,4}){1,7}|:)|fe80:(?::[0-9a-fA-F]{0,4}){0,4}%[0-9a-zA-Z]{1,}|::(?:ffff(?::0{1,4}){0,1}:){0,1}(?:(?:25[0-5]|(?:2[0-4]|1{0,1}[0-9]){0,1}[0-9])\.){3,3}(?:25[0-5]|(?:2[0-4]|1{0,1}[0-9]){0,1}[0-9])|(?:[0-9a-fA-F]{1,4}:){1,4}:(?:(?:25[0-5]|(?:2[0-4]|1{0,1}[0-9]){0,1}[0-9])\.){3,3}(?:25[0-5]|(?:2[0-4]|1{0,1}[0-9]){0,1}[0-9])))(?::[0-9]{1,4}|[1-5][0-9]{4}|6[0-4][0-9]{3}|65[0-4][0-9]{2}|655[0-2][0-9]|6553[0-5])?(?:/[\w\.-]*)*/?)\b"
+
+BASE_SCHEMA = {
+    "article_title": "",
+    "document": "",
+    "summary": "",
+    "keywords": [],
+    "authors": [],
+    "tags": [],
+    "meta_title": "",
+    "hashes": "",
+    "urls_from_document": [],
+    "url": "",
+}
+
+DATAFRAME_SCHEMA = {
+    "article_title": "",
+    "document": "",
+    "summary": "",
+    "meta_title": "",
+    "hashes": "",
+    "url": "",
+}
```

Comment:
maybe we make a named return type for these?

Reply:
looking further, we convert to pandas, at which point an apache arrow schema may make sense for dynamic typing

(trick: convert pd df to arrow, get arrow schema, then bake that generated arrow schema into code)